crates/analyzer/src/traversal/expressions.rs

```diff
@@ -705,18 +705,34 @@ fn expr_unary_operation(
 
         return match op.kind {
             fe::UnaryOperator::USub => {
-                let int_type = expected_type.as_int().unwrap_or(Integer::I256);
+                let expected_int_type = expected_type.as_int().unwrap_or(Integer::I256);
                 match operand_attributes.typ {
-                    Type::Base(Base::Numeric(_)) => {
+                    Type::Base(Base::Numeric(operand_type)) => {
                         if let fe::Expr::Num(num_str) = &operand.kind {
                             let num = -to_bigint(num_str);
-                            validate_numeric_literal_fits_type(context, num, exp.span, int_type);
+                            validate_numeric_literal_fits_type(
+                                context,
+                                num,
+                                exp.span,
+                                expected_int_type,
+                            );
+                        }
+
+                        if !(operand_type.is_signed() || expected_int_type.is_signed()) {
+                            context.error(
+                                "types always mismatch",
+                                operand.span,
+                                &format!(
+                                    "this has type `{}`; always return signed integer after apply '-' operator",
```

Comment:
minor grammer nitpicks: "this has type `{}`; always return**s** signed integer after apply**ing** '-' operator"